core/src/authorization/guards/action.guard.ts

```diff
@@ -0,0 +1,43 @@
+import type {CanActivate, ExecutionContext} from "@nestjs/common";
+import {Injectable} from "@nestjs/common";
+import {Reflector} from "@nestjs/core";
+import {GqlExecutionContext} from "@nestjs/graphql";
+
+import {MemberRepository} from "$repositories";
+
+import {JwtAuthPayloadSchema} from "../../authentication/types";
+import {AuthorizationService} from "../authorization.service";
+import {MetadataKeys} from "../authorization.types";
+
+@Injectable()
+export class ActionGuard implements CanActivate {
+    constructor(
+        private readonly reflector: Reflector,
+        private readonly authorizationService: AuthorizationService,
+        private readonly memberRepository: MemberRepository,
+    ) {}
+
+    async canActivate(context: ExecutionContext): Promise<boolean> {
+        const requiredActions = this.reflector.getAllAndOverride<string[] | undefined>(MetadataKeys.Actions, [
+            context.getHandler(),
+            context.getClass(),
+        ]);
+
+        if (!requiredActions?.length) return true;
+
+        const ctx = GqlExecutionContext.create(context).getContext();
+        if (!ctx?.req.user) return false;
+
+        const data = JwtAuthPayloadSchema.safeParse(ctx.req.user);
+        if (!data.success || !data.data.currentOrganizationId) return false;
```

Comment:
We might want a log here if we can't parse the JWT. That shouldn't normally happen, right?

Reply:
Right. Will do.